### PR TITLE
Homework 3: Henning Richter

### PR DESCRIPTION
**Pull request by hediriro**

### CODE REVIEW

File: src/main/java/impl/queue/ArrayQueueFactory.java [hunk 1]
```diff
@@ -2,8 +2,71 @@
 
 import api.queue.IntQueue;
 
-public class ArrayQueueFactory {
+public class ArrayQueueFactory implements IntQueue {
+    private static int capacity;
```
Comment (Walingar) [inline src/main/java/impl/queue/ArrayQueueFactory.java]: You use static fields here, so you can't create more than one ArrayIntQueue :(

File: src/main/java/impl/queue/ArrayQueueFactory.java [hunk 2]
```diff
@@ -2,8 +2,71 @@
 
 import api.queue.IntQueue;
 
-public class ArrayQueueFactory {
+public class ArrayQueueFactory implements IntQueue {
```
Comment (Walingar) [inline src/main/java/impl/queue/ArrayQueueFactory.java]: Factory shouldn't implement this interface. You need to create another class like `ArrayIntQueue` and return it's instance in getInstance method

File: src/main/java/impl/queue/ArrayQueueFactory.java [hunk 3]
```diff
@@ -2,8 +2,71 @@
...
+    private static int capacity;
+    private static Integer[] queueArray;
+    private static int currentSize;
```
Comment (Walingar) [inline src/main/java/impl/queue/ArrayQueueFactory.java]: You don't use this field

File: src/main/java/impl/queue/ArrayQueueFactory.java [hunk 4]
```diff
@@ -2,8 +2,71 @@
...
+        int currentSize = getSize();
+        if (currentSize == 0) {
+            return null;
+        } else {
+            return queueArray[0];
+        }
+    }
+
+    @Override
+    public int getSize() {
```
Comment (Walingar) [inline src/main/java/impl/queue/ArrayQueueFactory.java]: It uses linear time, but should use constant time

File: src/main/java/impl/queue/ArrayQueueFactory.java [hunk 5]
```diff
@@ -2,8 +2,71 @@
...
+    }
+
+    @Override
+    public Integer remove() {
+        int currentSize = getSize();
+        if (currentSize == 0) {
+            return null;
+        } else {
+            Integer output = queueArray[0];
+            for (int newPosition = 0; newPosition < capacity; newPosition++) {
```
Comment (Walingar) [inline src/main/java/impl/queue/ArrayQueueFactory.java]: It uses linear time but should use contant time

File: src/main/java/impl/queue/LinkedQueueFactory.java [hunk 6]
```diff
@@ -2,8 +2,113 @@
 
 import api.queue.IntQueue;
 
-public class LinkedQueueFactory {
+import java.util.LinkedList;
+
+class QueueNode {
```
Comment (Walingar) [inline src/main/java/impl/queue/LinkedQueueFactory.java]: Java file should contain only one class

File: src/main/java/impl/queue/LinkedQueueFactory.java [hunk 7]
```diff
@@ -2,8 +2,113 @@
...
+    public QueueNode getPrevious() {
+        return previous;
+    }
+
+    public void setPrevious(QueueNode previous) {
+        this.previous = previous;
+    }
+}
+
+public class LinkedQueueFactory implements IntQueue {
```
Comment (Walingar) [inline src/main/java/impl/queue/LinkedQueueFactory.java]: The same mistake as with ArrayQueueFactory.

File: src/main/java/impl/queue/LinkedQueueFactory.java (hunk 6 above)
Comment (Walingar) [inline src/main/java/impl/queue/LinkedQueueFactory.java]: You can move this class inside another class. You can read more here https://docs.oracle.com/javase/tutorial/java/javaOO/nested.html

File: src/main/java/impl/queue/LinkedQueueFactory.java [hunk 8]
```diff
@@ -2,8 +2,113 @@
...
+
+    public QueueNode getNext() {
+        return next;
+    }
+
+    public void setNext(QueueNode next) {
+        this.next = next;
+    }
+
+    public QueueNode getPrevious() {
```
Comment (Walingar) [inline src/main/java/impl/queue/LinkedQueueFactory.java]: You don't use this method

File: src/main/java/impl/queue/LinkedQueueFactory.java [hunk 9]
```diff
@@ -2,8 +2,113 @@
...
+        return previous;
+    }
+
+    public void setPrevious(QueueNode previous) {
```
Comment (Walingar) [inline src/main/java/impl/queue/LinkedQueueFactory.java]: You sets previous but don't get it

File: src/main/java/impl/queue/LinkedQueueFactory.java [hunk 10]
```diff
@@ -2,8 +2,113 @@
...
+        this.previous = previous;
+    }
+}
+
+public class LinkedQueueFactory implements IntQueue {
+    private static int capacity;
+    private static QueueNode queueLinked;
```
Comment (Walingar) [inline src/main/java/impl/queue/LinkedQueueFactory.java]: queueLinked -> linkedQueue as in english

File: src/main/java/impl/queue/LinkedQueueFactory.java [hunk 11]
```diff
@@ -2,8 +2,113 @@
...
+    public void add(int e) {
+        int currentSize = getSize();
+        if (currentSize == capacity) {
+            throw new IllegalStateException();
+        }
+        if (currentSize == 0) {
+            queueLinked.setValue(e);
+        } else {
+            QueueNode assist = queueLinked;
+            while (assist.getNext() != null) {
```
Comment (Walingar) [inline src/main/java/impl/queue/LinkedQueueFactory.java]: it takes linear time but shoud take constant time

File: src/main/java/impl/queue/LinkedQueueFactory.java [hunk 12]
```diff
@@ -2,8 +2,113 @@
...
+    public Integer remove() {
+        int currentSize = getSize();
+        if (currentSize == 0) {
+            return null;
+        } else {
+            Integer output = queueLinked.getValue();
+            if (queueLinked.getNext() == null) {
+                queueLinked.setValue(null);
+            } else {
+                queueLinked = queueLinked.getNext();
```
Comment (Walingar) [inline src/main/java/impl/queue/LinkedQueueFactory.java]: You are calling queueLinked.getNext() twice

File: src/main/java/impl/queue/LinkedQueueFactory.java [hunk 13]
```diff
@@ -2,8 +2,113 @@
...
+                queueLinked.setPrevious(null);
+            }
+            return output;
+        }
+    }
+
+    @Override
+    public Integer element() {
+        int currentSize = getSize();
+        if (currentSize == 0) {
```
Comment (Walingar) [inline src/main/java/impl/queue/LinkedQueueFactory.java]: you can inline this variable here: if (getSize() == 0) {...}

File: src/main/java/impl/queue/LinkedQueueFactory.java [hunk 14]
```diff
@@ -2,8 +2,113 @@
...
+
+    @Override
+    public int getSize() {
+        if (queueLinked.getValue() == null) {
+            return 0;
+        } else {
+            int size = 1;
+            QueueNode assist = queueLinked;
+            while (assist.getNext() != null) {
+                size += 1;
```
Comment (Walingar) [inline src/main/java/impl/queue/LinkedQueueFactory.java]: size++

File: src/main/java/impl/queue/LinkedQueueFactory.java [hunk 15]
```diff
@@ -2,8 +2,113 @@
...
+        int currentSize = getSize();
+        if (currentSize == 0) {
+            return null;
+        } else {
+            return queueLinked.getValue();
+        }
+    }
+
+    @Override
+    public int getSize() {
```
Comment (Walingar) [inline src/main/java/impl/queue/LinkedQueueFactory.java]: This method takes linear time, but should take constant time

**Review by Walingar — CHANGES_REQUESTED** [review]

File: src/main/java/impl/queue/LinkedQueueNode.java [hunk 16]
```diff
@@ -0,0 +1,37 @@
+package impl.queue;
+
+public class LinkedQueueNode {
```
Comment (Walingar) [inline src/main/java/impl/queue/LinkedQueueNode.java]: it's better to make it nested

File: src/main/java/impl/queue/ArrayQueueImpl.java [hunk 17]
```diff
@@ -0,0 +1,59 @@
+package impl.queue;
+
+import api.queue.IntQueue;
+
+public class ArrayQueueImpl implements IntQueue {
+    private final int capacity;
+    private final Integer[] queueArray;
+    private int currentSize;
+    private int currentShift;
```
Comment (Walingar) [inline src/main/java/impl/queue/ArrayQueueImpl.java]: I guess it is better to rename it to "headIndex", isn't it?

File: src/main/java/impl/queue/ArrayQueueImpl.java [hunk 18]
```diff
@@ -0,0 +1,59 @@
+package impl.queue;
+
+import api.queue.IntQueue;
+
+public class ArrayQueueImpl implements IntQueue {
+    private final int capacity;
+    private final Integer[] queueArray;
+    private int currentSize;
```
Comment (Walingar) [inline src/main/java/impl/queue/ArrayQueueImpl.java]: just `size` `current` is redundant

File: src/main/java/impl/queue/ArrayQueueImpl.java [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+        currentShift = 0;
+    }
+
+    @Override
+    public void add(int e) {
+        if (currentSize == capacity) {
+            throw new IllegalStateException();
+        } else {
+            int currentIndex = currentSize + currentShift;
+            if (currentIndex >= capacity) {
```
Comment (Walingar) [inline src/main/java/impl/queue/ArrayQueueImpl.java]: Can we use `%` here?

**Review by Walingar — CHANGES_REQUESTED** [review]

File: src/main/java/impl/queue/LinkedQueueImpl.java [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+        }
+        return head.getValue();
+    }
+
+    @Override
+    public int getSize() {
+        return currentSize;
+    }
+
+    class LinkedQueueNode {
```
Comment (Walingar) [inline src/main/java/impl/queue/LinkedQueueImpl.java]: You can mark it as static

File: src/main/java/impl/queue/ArrayQueueImpl.java [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+    public ArrayQueueImpl(int maxSize) {
+        capacity = maxSize;
+        queueArray = new Integer[capacity];
+        size = 0;
+        headIndex = 0;
+    }
+
+    @Override
+    public void add(int e) {
+        if (size == capacity) {
```
Comment (Walingar) [inline src/main/java/impl/queue/ArrayQueueImpl.java]: ArrayQueueImpl and LinkedQueueImpl contain similar part of code like this check, so extract it to an abstract class. Use all OOP power!

**Review by Walingar — CHANGES_REQUESTED** [review]

File: src/main/java/impl/queue/ArrayQueueImpl.java (hunk 21 above)
Comment (hediriro) [reply src/main/java/impl/queue/ArrayQueueImpl.java]: Actually, I don't think that an abstract class would improve this code here. There are only few similarites,. The most are in if conditions where the else block differs tremendously. I don't think that I could just overwrite the else block for each instanciation while keeping the code not that complex. If I applied that, the complexity increased extraordinarily, as I had to write many new methods called in any affected else block. Thus, an abstract class would only make sense for getSize(). But creating an abstract class just for saving one method so incredibly small doesn't seem to be worth the effort from my point of view.

**Comment by hediriro** [comment]
Sorry for changing somtething again before re-requesting your review, but I thought afterwards that removing these unnecessary getters and setters might be senseful.

File: src/main/java/impl/queue/LinkedQueueImpl.java [hunk 22]
```diff
@@ -0,0 +1,61 @@
+package impl.queue;
+
+import api.queue.IntQueue;
```
Comment (Walingar) [inline src/main/java/impl/queue/LinkedQueueImpl.java]: redundant import

**Review by Walingar — CHANGES_REQUESTED** [review]

**Review by Walingar — APPROVED** [review]

**Pull request merged**